From: Steffen Maier <maier@linux.vnet.ibm.com>
Date: Wed, 10 Aug 2016 18:30:44 +0200
Subject: zfcp: fix fc_host port_type with NPIV
Git-commit: bd77befa5bcff8c51613de271913639edf85fbc2
Patch-mainline: v4.9-rc1
References: bnc#958000, LTC#132479

Description:  zfcp: fix fc_host port_type with NPIV
Symptom:      For an NPIV-enabled FCP device, zfcp can erroneously show
              "NPort (fabric via point-to-point)" instead of "NPIV VPORT"
              for the port_type sysfs attribute of the corresponding
              fc_host.
              s390-tools that can be affected are dbginfo.sh and ziomon.
Problem:      zfcp_fsf_exchange_config_evaluate() ignores
              fsf_qtcb_bottom_config.connection_features indicating NPIV
              and only sets fc_host_port_type to FC_PORTTYPE_NPORT if
              fsf_qtcb_bottom_config.fc_topology is FSF_TOPO_FABRIC.
              Only the independent zfcp_fsf_exchange_port_evaluate()
              evaluates connection_features to overwrite fc_host_port_type
              to FC_PORTTYPE_NPIV in case of NPIV.
              Code was introduced with upstream kernel 2.6.30
              commit 0282985da5923fa6365adcc1a1586ae0c13c1617
              ("[SCSI] zfcp: Report fc_host_port_type as NPIV").
              This works during FCP device recovery (such as set online)
              because it performs FSF_QTCB_EXCHANGE_CONFIG_DATA followed by
              FSF_QTCB_EXCHANGE_PORT_DATA in sequence.
              However, the zfcp-specific scsi host sysfs attributes
              "requests", "megabytes", or "seconds_active" trigger only
              zfcp_fsf_exchange_config_evaluate() resetting fc_host
              port_type to FC_PORTTYPE_NPORT despite NPIV.
              The zfcp-specific scsi host sysfs attribute "utilization"
              triggers only zfcp_fsf_exchange_port_evaluate() correcting
              the fc_host port_type again in case of NPIV.
Solution:     Evaluate fsf_qtcb_bottom_config.connection_features
              in zfcp_fsf_exchange_config_evaluate() where it belongs to.
Reproduction: Read any of the following zfcp-specific scsi host sysfs
              attributes "requests", "megabytes", or "seconds_active" in
              /sys/devices/css*/*.*.*/*.*.*/host*/scsi_host/host*/ .
              After that, despite NPIV, fc_host port_type shows
              "NPort (fabric via point-to-point)" until any of the
              following happens:
              FCP device recovery (no matter if automatic to recover some
              error or if manually user triggered such as: write 0 to
              failed attribute, offline&online, CHPID off&on, z/VM CP
              DETACH&ATTACH), or
              read the zfcp-specific scsi host sysfs attribute
              "utilization" in
              /sys/devices/css*/*.*.*/*.*.*/host*/scsi_host/host*/ .


Signed-off-by: Steffen Maier <maier@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/scsi/zfcp_fsf.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

--- a/drivers/s390/scsi/zfcp_fsf.c
+++ b/drivers/s390/scsi/zfcp_fsf.c
@@ -513,7 +513,10 @@ static int zfcp_fsf_exchange_config_eval
 		fc_host_port_type(shost) = FC_PORTTYPE_PTP;
 		break;
 	case FSF_TOPO_FABRIC:
-		fc_host_port_type(shost) = FC_PORTTYPE_NPORT;
+		if (bottom->connection_features & FSF_FEATURE_NPIV_MODE)
+			fc_host_port_type(shost) = FC_PORTTYPE_NPIV;
+		else
+			fc_host_port_type(shost) = FC_PORTTYPE_NPORT;
 		break;
 	case FSF_TOPO_AL:
 		fc_host_port_type(shost) = FC_PORTTYPE_NLPORT;
@@ -618,7 +621,6 @@ static void zfcp_fsf_exchange_port_evalu
 
 	if (adapter->connection_features & FSF_FEATURE_NPIV_MODE) {
 		fc_host_permanent_port_name(shost) = bottom->wwpn;
-		fc_host_port_type(shost) = FC_PORTTYPE_NPIV;
 	} else
 		fc_host_permanent_port_name(shost) = fc_host_port_name(shost);
 	fc_host_maxframe_size(shost) = bottom->maximum_frame_size;
